File: sem1/hw7/task1/tree.cpp
#include "tree.h"
#include <iostream>

using namespace std;

struct TreeNode
{
   int value;
   TreeNode *leftChild;
   TreeNode *rightChild;
};

struct Tree
{
    TreeNode *root;
};

Tree *createTree()
{
    Tree *tree = new Tree;
    tree->root = nullptr;
    return tree;
}

TreeNode *createNewNode(int value, TreeNode *left, TreeNode *right)
{
    TreeNode *node = new TreeNode;
    node->value = value;
    node->leftChild = left;
    node->rightChild = right;
    return node;
}

void deleteTree(TreeNode *node)
{
    if (node == nullptr)
        return;

     deleteTree(node->leftChild);
     deleteTree(node->rightChild);
     delete node;
}

void deleteTree(Tree *tree)
{
    deleteTree(tree->root);
    delete tree;
}

void printAscendingTree(TreeNode *node)
{
    if (node == nullptr)
        return;

    printAscendingTree(node->leftChild);
    cout << node->value << " ";
    printAscendingTree(node->rightChild);
}

void printAscendingTree(Tree *tree)
{
    printAscendingTree(tree->root);
}

void printDescendingTree(TreeNode *node)
{
    if (node == nullptr)
        return;

    printDescendingTree(node->rightChild);
    cout << node->value << " ";
    printDescendingTree(node->leftChild);
}

void printDescendingTree(Tree *tree)
{
    printDescendingTree(tree->root);
}

void printTree(TreeNode *node)
{
    if (node == nullptr)
    {
        cout << " null";
        return;
    }

    cout << " (";
    cout << node->value;
    printTree(node->leftChild);
    printTree(node->rightChild);
    cout << ")";
}

void printTree(Tree *tree)
{
    printTree(tree->root);
}

void add(TreeNode *&node, int value)
{
    if (node == nullptr)
    {
        node = createNewNode(value, nullptr, nullptr);
        return;
    }
    if (node->value == value)
        return;
    if (node->value > value)
        add(node->leftChild, value);
    if (node->value < value)
        add(node->rightChild, value);
}

void add(Tree *tree, int value)
{
    add(tree->root, value);
}

bool hasNoChildren(TreeNode *node)
{
    return (node->leftChild == nullptr && node->rightChild == nullptr);
}

bool hasOnlyLeftChild(TreeNode *node)
{
    return (node->leftChild != nullptr && node->rightChild == nullptr);
}

bool hasOnlyRightChild(TreeNode *node)
{
    return (node->leftChild == nullptr && node->rightChild != nullptr);
}

bool hasTwoChildren(TreeNode *node)
{
    return (node->leftChild != nullptr && node->rightChild != nullptr);
}

bool remove(TreeNode *&node, int value)
{
    if (node == nullptr)
        return false;
    if (value == node->value)
    {
        if (hasNoChildren(node))
        {
            delete node;
            node = nullptr;
        }
        else if (hasOnlyLeftChild(node))
        {
            TreeNode *toDelete = node;
            node = node->leftChild;
            delete toDelete;
        }
        else if (hasOnlyRightChild(node))
        {
            TreeNode *toDelete = node;
            node = node->rightChild;
            delete toDelete;
        }
        else if (hasTwoChildren(node))
        {
            TreeNode *temp = node->rightChild;
            while (temp->leftChild != nullptr)
                temp = temp->leftChild;
            int a = temp->value;
            remove(node, a);
            node->value = a;
           // delete temp;
        }
        return true;
    }
    if (value > node->value)
        return remove(node->rightChild, value);
    if (value < node->value)
        return remove(node->leftChild, value);
}

void remove(Tree *tree, int value)
{
    remove(tree->root, value);
}

bool exists(TreeNode *node, int value)
{
    if (node == nullptr)
        return false;
    if (value == node->value)
        return true;
    if (value > node->value)
        return exists(node->rightChild, value);
    if (value < node->value)
        return exists(node->leftChild, value);
}

bool exists(Tree *tree, int value)
{
    return exists(tree->root, value);
}

File: sem3/hw2/task2/src/spbu/sem3/hw2/task2/Tree.java
package spbu.sem3.hw2.task2;

import java.util.Iterator;
import java.util.NoSuchElementException;
import java.util.Stack;

/** class for Binary Search Tree with Iterator implementation. */
public class Tree<T extends Comparable> implements Iterable<T>{
    private Node<T> root = null;

    /* method that adds some value to tree if it hasn't stored there already. */
    public void add(T value) {
        root = add(value, root);
    }

    /* method that adds some value to tree that has given node as its root. */
    private Node add(T value, Node<T> node) {
        if (node == null) {
            node = new Node(value, null, null);
            return node;
        }
        int sign = node.value.compareTo(value);

        if (sign == 0) {
            return node;
        } else if (sign > 0) {
            node.left = add(value, node.left);
        } else {
            node.right = add(value, node.right);
        }
        return node;
    }

    /* method that removes value from tree if it exists there. */
    public void remove(T value) {
        root = remove(value, root);
    }

    /* method that removes value from tree with given node as its root. */
    private Node remove(T value, Node<T> node) {
        if (node == null) {
            return node;
        }
        int sign = node.value.compareTo(value);
        if (sign > 0) {
            node.left = remove(value, node.left);
        } else if (sign < 0) {
            node.right = remove(value, node.right);
        } else {
            if (node.hasNoChildren()) {
                node = null;
            } else if (node.hasOnlyLeftChild()) {
                node = node.left;
            } else if (node.hasOnlyRightChild()) {
                node = node.right;
            } else {
                Node temp = node.right;
                while (temp.left != null) {
                    temp = temp.left;
                }
                T data = (T)temp.value;
                remove(data);
                node.value = data;
            }
        }
        return node;
    }

    /* method that returns true if tree is empty and false otherwise. */
    public boolean isEmpty() {
        return root == null;
    }

    public String printTree() {
        return printTree(root);
    }

    private String printTree(Node<T> node) {
        if (node == null)
            return "";
        StringBuilder sb = new StringBuilder();
        sb.append(" (");
        sb.append(node.value);
        sb.append(printTree(node.left));
        sb.append(printTree(node.right));
        sb.append(")");
        return sb.toString();
    }

    /* method that generates iterator for tree. */
    @Override
    public Iterator<T> iterator() {
        return new TreeIterator(root);
    }

    /* class for BST Element - Node. */
    private class Node<T extends Comparable> {
        private T value;
        private Node<T> left;
        private Node<T> right;

        private Node(T value, Node left, Node right) {
            this.value = value;
            this.left = left;
            this.right = right;
        }

        private boolean hasNoChildren() {
            return left == null && right == null;
        }
        private boolean hasOnlyRightChild() {
            return left == null && right != null;
        }
        private boolean hasOnlyLeftChild() {
            return left != null && right == null;
        }
    }

    /* class for tree iterator. */
    private class TreeIterator implements Iterator<T> {
        Stack<Node> stack;

        /* creation of iterator. */
        public TreeIterator(Node root) {
            stack = new Stack<>();
            while (root != null) {
                stack.push(root);
                root = root.left;
            }
        }

        /* method that returns false if there is anything else in tree and true otherwise. */
        @Override
        public boolean hasNext() {
            return !stack.isEmpty();
        }

        /* method that return next element of tree (in in-order traversal) or throws exception */
        @Override
        public T next() {
            if (!hasNext()) {
                throw new NoSuchElementException();
            }
            Node next = stack.pop();
            Node newRoot = next.right == null ? null : next.right;
            while (newRoot != null) {
                stack.push(newRoot);
                newRoot = newRoot.left;
            }
            return (T)next.value;
        }
    }
}
